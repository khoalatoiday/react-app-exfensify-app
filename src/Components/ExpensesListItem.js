import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
// sử dụng với expense={expense}
// const Item = (props) => (
//     <div>
//         <p>{props.expense.description}</p>
//         <p>{props.expense.createAt} - {props.expense.amount}</p>
//     </div>
// )

// sử dụng với {...expense} -> khi đó props của Item chính là Object expense
const Item = ({ id, description, amount, createAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {moment(createAt).format("dddd, MMMM Do YYYY, h:mm:ss a")} -{" "}
      {numeral(amount / 100).format("$0,0.00")}
    </p>
  </div>
);

export default Item;
