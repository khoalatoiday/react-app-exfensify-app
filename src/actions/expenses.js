const uuid = require("uuid");

const ADD_EXPENSES = ({
  description = "",
  note = "",
  amount = 0,
  createAt = 0,
} = {}) => ({
  type: "ADD_EXPENSES",
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createAt,
  },
});

const REMOVE_EXPENSES = ({ id } = {}) => ({
  type: "REMOVE_EXPENSES",
  expense: {
    id,
  },
});

const EDIT_EXPENSE = (id, updates) => ({
  // updates là một Object
  type: "EDIT_EXPENSES",
  id,
  updates,
});

export {ADD_EXPENSES, REMOVE_EXPENSES, EDIT_EXPENSE}