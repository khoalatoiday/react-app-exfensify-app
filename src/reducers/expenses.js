//Expenses Reducer
const expensesDefaultValueState = [];
// tại node đầu tiên thì không có state -> set up state = DefaultState
export default (state = expensesDefaultValueState, action) => {
  switch (action.type) {
    case "ADD_EXPENSES":
      return [...state, action.expense]; // new array với toàn bộ phần tử trong state + action.expense
    case "REMOVE_EXPENSES":
      return state.filter((expense) => action.expense.id != expense.id); // trả về một dãy thỏa mãn điều kiện
    case "EDIT_EXPENSES":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        }else{
          return {
            ...expense
          }
        }
      });
    default:
      return state;
  }
};
