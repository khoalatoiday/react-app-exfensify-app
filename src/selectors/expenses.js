import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  // retrun a new array thỏa mãn các điều kiện
  return expenses
    .filter((expense) => {
      const createAtMoment = moment(expense.createAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createAtMoment, "day")
        : true; // moment query doc, startDate có <= thời gian được tạo của expense
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createAtMoment, "day")
        : true; // endDate có >= thời gian được tạo của expense
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((firstEle, secondEle) => {
      if (sortBy === "date") {
        return firstEle.createAt < secondEle.createAt ? 1 : -1; // nếu xảy ra điều kiện thì đẩy secondEle lên trước -> sắp xếp giảm dần
      }
      if (sortBy === "amount") {
        return firstEle.amount > secondEle.amount ? 1 : -1; // nếu xảy ra điều kiện thì đẩy secondEle lên trước -> sắp xếp tăng dần
      }
    });
};

export { getVisibleExpenses as default };
