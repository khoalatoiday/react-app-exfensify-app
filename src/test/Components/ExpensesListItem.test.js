import React from "react";
import { shallow } from "enzyme";
import ExpensesListItem from "../../Components/ExpensesListItem"
import expenses from "../fixtures/expense"

test("Should render Expenese List Item",()=>{
    const wrapper = shallow(<ExpensesListItem expense = {expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})