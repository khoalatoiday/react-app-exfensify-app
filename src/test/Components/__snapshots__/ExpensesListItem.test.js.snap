// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expenese List Item 1`] = `
<div>
  <Link
    to="/edit/undefined"
  >
    <h3 />
  </Link>
  <p>
    Thursday, January 1st 1970, 8:00:00 am
     -
     
    $0.00
  </p>
</div>
`;
